File: Tabuleiro.java
public class Tabuleiro {
    private int[][] table;
    private int length;

    public Tabuleiro(String[] numbers){
        length = (int) Math.sqrt(numbers.length);
        table = new int[length][length];
        int k= 0;
        for(int i = 0; i < length; i++){
            for(int j=0; j<length; j++){
                table[i][j] = Integer.parseInt(numbers[k++]);
            }
        }
    }

    public void UpMove(){
        for(int i = 0; i < length; i++){
            for(int j=0; j<length; j++){
                if(table[i][j] == 0 && i < length-1){
                    int temp = table[i][j];
                    table[i][j] = table[i+1][j];
                    table[i+1][j] = temp;
                    return;
                }
            }
        }
    }
    public void DownMove(){
        for(int i = 0; i < length; i++){
            for(int j=0; j<length; j++){
                if(table[i][j] == 0 && i > 0){
                    int temp = table[i][j];
                    table[i][j] = table[i-1][j];
                    table[i-1][j] = temp;
                    return;
                }
            }
        }
    }
    public void LeftMove(){
        for(int i = 0; i < length; i++){
            for(int j=0; j<length; j++){
                if(table[i][j] == 0 && j < length-1){
                    int temp = table[i][j];
                    table[i][j] = table[i][j+1];
                    table[i][j+1] = temp;
                    return;
                }
            }
        }
    }
    public void RightMove(){
        for(int i = 0; i < length; i++){
            for(int j=0; j<length; j++){
                if(table[i][j] == 0 && j > 0){
                    int temp = table[i][j];
                    table[i][j] = table[i][j-1];
                    table[i][j-1] = temp;
                    return;
                }
            }
        }
    }

    public void PrintTabuleiro(){
        for(int i = 0; i < length; i++){
            for(int j = 0; j < length; j++){
                System.out.print("+------");
            }
            System.out.println("+");
            for(int j=0; j<length; j++){
                if(table[i][j] < 10){
                    System.out.print("|   ");
                }else{
                    System.out.print("|  ");
                }
                System.out.print(table[i][j]+"  ");
                if(j == length-1){
                    System.out.print("|");
                }
            }
            System.out.print("\n");
        }
        for(int j = 0; j < length; j++){
            System.out.print("+------");
        }
        System.out.println("+");
        System.out.println("");
    }

    public boolean Solucao(){
        int cont = 0;
        for(int i = 0; i < length; i++){
            for(int j=0; j<length; j++){
                if(table[i][j]!=cont++){
                    return false;
                }
            }
        }
        return true;
    }
}

File: tabuleiroPython/tabuleiro.py
from math import sqrt

class Tabuleiro:
    def __init__(self, numbers):
        self.length = int(sqrt(len(numbers)))

        self.table = [[0 for _ in range(self.length)] for _ in range(self.length)]

        k = 0
        for i in range(self.length):
            for j in range(self.length):
                self.table[i][j] = int(numbers[k])
                k += 1

    def up(self):
        for i in range(self.length):
            for j in range(self.length):
                if(self.table[i][j] == 0 and i < self.length-1):
                    temp = self.table[i][j]
                    self.table[i][j] = self.table[i+1][j]
                    self.table[i+1][j] = temp
                    return
    
    def down(self):
        for i in range(self.length):
            for j in range(self.length):
                if(self.table[i][j] == 0 and i > 0):
                    temp = self.table[i][j]
                    self.table[i][j] = self.table[i-1][j]
                    self.table[i-1][j] = temp
                    return

    def left(self):
        for i in range(self.length):
            for j in range(self.length):
                if(self.table[i][j] == 0 and j < self.length-1):
                    temp = self.table[i][j]
                    self.table[i][j] = self.table[i][j+1]
                    self.table[i][j+1] = temp
                    return

    def right(self):
        for i in range(self.length):
            for j in range(self.length):
                if(self.table[i][j] == 0 and j > 0):
                    temp = self.table[i][j]
                    self.table[i][j] = self.table[i][j-1]
                    self.table[i][j-1] = temp
                    return

    def solucao(self):
        cont = 0
        for i in range(self.length):
            for j in range(self.length):
                if(self.table[i][j] == cont):
                    cont+=1
                else:
                    return False
        return True

    def printTab(self):
        for i in range(self.length):
            for j in range(self.length):
                print("+------",end="")
            print("+\n",end="|")
            for j in range(self.length):
                num = " " if self.table[i][j] == 0 else self.table[i][j]
                print(f"{num:>4}",end="  |")
            print("")
        for j in range(self.length):
            print("+------",end="")
        print("+\n")